File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    ArrayList<Integer>[] list;
    boolean[] visited;
   
    public int solution(int n, int[][] wires) {
        int answer = 101;
       
        list = new ArrayList[n+1];
        
        for(int i = 1; i <=n; i++) {
            list[i] = new ArrayList<>();
        }
        for(int i = 0; i < wires.length; i++) {
            int u = wires[i][0];
            int v = wires[i][1];
            list[u].add(v);
            list[v].add(u);
        }
        for(int i = 0; i < wires.length; i++) {
            visited = new boolean[n+1];
            int u = wires[i][0];
            int v = wires[i][1];
            list[u].remove((Integer) v);  
            list[v].remove((Integer) u);
            int[] size = new int[2];
            int index = 0;
            for(int j = 1; j <=n; j++) {
                if(!visited[j]) {
                    size[index++] = dfs(j); 
                }
            }
            answer = Math.min(answer,Math.abs(size[1]-size[0]));
            list[u].add(v);
            list[v].add(u);
            
        }
        
        return answer;
    }
    int dfs(int v) {
        int count = 0;
        visited[v] = true;
        for(int i : list[v]) {
            if(!visited[i]) {
                visited[i] = true;
                count += dfs(i);
                count++;
            }
        }
    return count;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function dfs(node, graph, visited) {
    visited.add(node);
    let count = 1;
    for (const next of graph[node]) {
        if(!visited.has(next)) {
            visited.add(next);
            count += dfs(next,graph,visited);
        }
    }
    return count;
}

function solution(n, wires) {
    var answer = Infinity;
    
    for (let i = 0; i < wires.length; i++) {
        const visited = new Set();
        const graph = {};
        for (let j = 1; j <= n; j++) graph[j] = [];
        for (let j = 0; j < wires.length; j++) {
            if (i === j) continue;
            const [u, v] = wires[j];
            graph[u].push(v);
            graph[v].push(u);
        }
        const first = dfs(1,graph,visited);
        const second = n - first;
        answer = Math.min(answer, Math.abs(first-second));
        
    }
    return answer;
}